### PR TITLE
CORE-19372 Use databases for cluster schema

### PR DESCRIPTION
**Pull request by davidcurrie**
This PR removes the concept of a "cluster DB". Instead, the `databases` stanza introduced for state managers can now also be used for the config, crypto, RBAC, and virtual node databases. By default, they will all share a database as before but they can be split apart. The other, perhaps more significant, change, is that the automatic bootstrap will now create separate users for each of crypto, RBAC, and virtual node databases, and one for worker that accesses the config DB. It is, therefore, now only necessary to give a single bootstrap user for each database and everything else is handled by the chart.

### CODE REVIEW

File: charts/corda/values.yaml [hunk 1]
```diff
@@ -403,39 +365,16 @@ bootstrap:
...
-      schema: CRYPTO
+      # -- persistent storage type
+      type: Database
+      # -- name of the persistent storage to use. When using 'Database' as the type, an element with the given id should exist under the databases section
+      storageId: "default"
+      # -- name of the partition to use. When using 'Database' as the type, the 'partition' represents the name of the database schema
+      partition: "CRYPTO"
       # JDBC connection pool configuration for crypto DB
-      dbConnectionPool:
+      connectionPool:
```
Comment (jujoramos) [inline charts/corda/values.yaml:353]: Do we need the `connectionPool` settings under `crypto` and `rbac` for `bootstrap`?. It’s only used for setting things up and only for development (`bootstrap` is not expected to be used in production environments, is it?), so I’d get rid of it entirely.

Comment (davidcurrie) [reply charts/corda/values.yaml:353]: These aren't the settings used at bootstrap time, they are the settings that bootstrap puts into the connections table for use at runtime.

Comment (jujoramos) [reply charts/corda/values.yaml:353]: Got it 👍

File: .ci/e2eTests/Jenkinsfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-@Library('corda-shared-build-pipeline-steps@5.2') _
```
Comment (davidcurrie) [inline .ci/e2eTests/Jenkinsfile:1]: Revert once pipeline changes are merged.

File: Jenkinsfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-@Library('corda-shared-build-pipeline-steps@5.2') _
```
Comment (davidcurrie) [inline Jenkinsfile:1]: Revert once pipeline changes are merged.

File: tools/plugins/preinstall/src/main/kotlin/net/corda/cli/plugins/preinstall/CheckPostgres.kt [hunk 4]
```diff
@@ -28,59 +28,61 @@ class CheckPostgres : Callable<Int>, PluginContext() {
...
+                    connect(postgresUrl, bootstrapUsername, bootstrapPassword, "bootstrap", db.id)
+                }
+            }
+        } else {
+            // We can only check runtime credentials if not using bootstrap as they won't have been created yet
+            val dbConfig = yaml.databases.first { it.id == yaml.config.storageId }
+            yaml.workers.forEach { (name, worker) ->
+                if (worker.config != null) {
+                    checkWorkerConnection(worker.config, name, dbConfig, yaml.config.storageId)
+                }
```
Comment (jujoramos) [inline tools/plugins/preinstall/src/main/kotlin/net/corda/cli/plugins/preinstall/CheckPostgres.kt]: Can we also check the connection for `stateTypes` defined under the `stateManager` section here?.

**Review by jujoramos — DISMISSED** [review]
Minor request, otherwise looks good to me 👍 .

File: tools/plugins/preinstall/src/main/kotlin/net/corda/cli/plugins/preinstall/CheckPostgres.kt (hunk 4 above)
Comment (davidcurrie) [reply tools/plugins/preinstall/src/main/kotlin/net/corda/cli/plugins/preinstall/CheckPostgres.kt]: yes - will do...

**Comment by jujoramos** [comment]
@davidcurrie : one thing I've noted and forgot to ask (not related to this `PR` but, since we're already modifying these files...): is there a reason for having `imagePullPolicy: Always` by default instead of `IfNotPresent`?.

**Comment by davidcurrie** [comment]
> @davidcurrie : one thing I've noted and forgot to ask (not related to this `PR` but, since we're already modifying these files...): is there a reason for having `imagePullPolicy: Always` by default instead of `IfNotPresent`?.

It's considered best practice as it prevents containers using images that are available on the node but which it doesn't actually have permission to pull (they were pulled on behalf of another container).

File: .ci/e2eTests/Jenkinsfile (hunk 2 above)
Comment (jujoramos) [reply .ci/e2eTests/Jenkinsfile:1]: In order to reduce disruption with other teams, this has to be checked in as it is and reverted afterwards. This is exactly what we did for the previous breaking Helm changes last week (with @ronanbrowne 's blessing).
The steps (just for my reference) are below:
- Merge `corda-runtime-os` with the pipelines pointing to `@cluster-db/CORE-19372` and verify things work as expected.
- Merge https://github.com/corda/corda-shared-build-pipeline-steps/pull/1597 (if GitHub deletes the branch, re-create it immediately).
- Open a `PR` against `corda-runtime-os` to updating the pipelines pointing back to `@5.2`and verify things work as expected.
- Merge `corda-runtime-os` `PR` and delete `cluster-db/CORE-19372` branch from `corda-shared-build-pipeline-steps` repository.

File: Jenkinsfile (hunk 3 above)
Comment (jujoramos) [reply Jenkinsfile:1]: Same as https://github.com/corda/corda-runtime-os/pull/5643#discussion_r1483325235.

**Review by jujoramos — APPROVED** [review]
LGTM

**Review by mag1c-48 — APPROVED** [review]
LGTM

**Pull request merged**